third_party/move/move-compiler-v2/src/pipeline/uninitialized_use_checker.rs

```diff
@@ -0,0 +1,386 @@
+};
+use std::collections::BTreeMap;
+
+/// State of initialization of a local at a given program point.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Initialized {
+    No,    // definitely not initialized
+    Maybe, // maybe initialized
+    Yes,   // definitely initialized
+}
+
+impl AbstractDomain for Initialized {
+    /// Implements `join` for the initialized state lattice:
+    ///
+    ///             +-------+
+    ///             | Maybe |
+    ///             +-------+
+    ///               /   \
+    ///              /     \
+    ///      +-----+/       \+----+
+    ///      | Yes |         | No |
+    ///      +-----+\       /+----+
+    ///              \     /
+    ///               \   /
+    ///            +--------+
+    ///            | bottom |
+    ///            +--------+
+    /// Note that bottom is not explicitly represented in the enum `Initialized`.
+    /// Instead, it is implicit: it represents the initialized state for locals at unreachable program points.
+    fn join(&mut self, other: &Self) -> JoinResult {
+        if *self == *other {
+            return JoinResult::Unchanged;
+        }
+        if *self != Initialized::Maybe {
+            *self = Initialized::Maybe;
+            JoinResult::Changed
+        } else {
+            JoinResult::Unchanged
+        }
+    }
+}
+
+/// Initialization state of all the locals at a program point.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct InitializedState(Vector<Initialized>);
+
+impl InitializedState {
+    /// Create a new initialized state with:
+    /// * all param locals set to `Yes`
+    /// * all other locals set to `No`.
+    /// Note: `num_locals` is the total number of locals, including params.
+    pub fn new(num_params: usize, num_locals: usize) -> Self {
+        if num_locals < num_params {
+            panic!("ICE: num_locals must be >= num_params");
+        }
+        Self(Vector::from_iter(
+            std::iter::repeat(Initialized::Yes)
+                .take(num_params)
+                .chain(std::iter::repeat(Initialized::No).take(num_locals - num_params)),
+        ))
+    }
+
+    /// Mark `local` as initialized.
+    fn mark_as_initialized(&mut self, local: usize) {
+        self.0.set(local, Initialized::Yes);
+    }
+
+    /// Get the initialization state of `local`.
+    /// Panics if `local` does not exist in this state.
+    fn get_initialized_state(&self, local: usize) -> Initialized {
+        self.0.get(local).expect("local must exist").clone()
+    }
+}
+
+impl AbstractDomain for InitializedState {
+    /// `join` of two initialized states is the point-wise join for each local.
+    /// The result is `JoinResult::Changed` if any of the local joins causes a change.
+    fn join(&mut self, other: &Self) -> JoinResult {
+        let mut result = JoinResult::Unchanged;
+        // Do a shallow check if the two states are the same.
+        if self.0.ptr_eq(&other.0) {
```

Comment:
Will this ever actually be the case?

Reply:
Yes, when other is a clone of self (without modifications). In the dataflow analysis framework, there is a lot of state cloning (for example, when propagating dataflow facts), so I believe this should help speed up the joins.